engines/cufile.c

```diff
@@ -0,0 +1,586 @@
+
+	if (fcd != NULL) {
+		cuFileHandleDeregister(fcd->cf_handle);
+		FILE_SET_ENG_DATA(f, NULL);
+		free(fcd);
+	}
+
+	rc = generic_close_file(td, f);
+
+	return rc;
+}
+
+static int fio_cufile_iomem_alloc(struct thread_data *td, size_t total_mem)
+{
+	struct cufile_options *o = td->eo;
+	int rc;
+	CUfileError_t status;
+
+	o->total_mem = total_mem;
+	o->logged = 0;
+	td->orig_buffer = calloc(1, total_mem);
+	if (!td->orig_buffer) {
+		log_err("calloc failed: %d\n", errno);
+		goto exit_error;
+	}
+
+	if (o->use_posixio) {
+		o->junk_buf = calloc(1, total_mem);
+		if (o->junk_buf == NULL) {
+			goto exit_error;
+		}
+	} else {
+		o->junk_buf = NULL;
+	}
+
+	dprint(FD_MEM, "Alloc %zu for GPU %d\n", total_mem, o->my_gpu_id);
+	check_cudaruntimecall(cudaMalloc(&o->cu_mem_ptr, total_mem), rc);
+	if (rc != 0) {
+		goto exit_error;
+	}
+	check_cudaruntimecall(cudaMemset(o->cu_mem_ptr, 0xab, total_mem), rc);
+	if (rc != 0) {
+		goto exit_error;
+	}
+
+	if (!o->use_posixio) {
+		status = cuFileBufRegister(o->cu_mem_ptr, total_mem, 0);
+		if (status.err != CU_FILE_SUCCESS) {
+			log_err("cuFileBufRegister: %d:%s\n", status.err,
+					cuFileGetErrorString(status));
+			goto exit_error;
+		}
+	}
+
+	return 0;
+
+exit_error:
+	if (td->orig_buffer) {
+		free(td->orig_buffer);
+		td->orig_buffer = NULL;
+	}
+	if (o->junk_buf) {
+		free(o->junk_buf);
+		o->junk_buf = NULL;
+	}
+	if (o->cu_mem_ptr) {
+		cudaFree(o->cu_mem_ptr);
+		o->cu_mem_ptr = NULL;
+	}
+	return 1;
+}
+
+static void fio_cufile_iomem_free(struct thread_data *td)
+{
+	struct cufile_options *o = td->eo;
+	if (o->junk_buf) {
+		free(o->junk_buf);
+		o->junk_buf = NULL;
+	}
+	if (o->cu_mem_ptr) {
+		if (!o->use_posixio) {
+			cuFileBufDeregister(o->cu_mem_ptr);
+		}
+		cudaFree(o->cu_mem_ptr);
+		o->cu_mem_ptr = NULL;
+	}
+	if (td->orig_buffer) {
+		free(td->orig_buffer);
+		td->orig_buffer = NULL;
+	}
+}
+
+static void fio_cufile_cleanup(struct thread_data *td)
+{
+	struct cufile_options *o = td->eo;
+	pthread_mutex_lock(&running_lock);
+	running--;
+	assert(running >= 0);
+	if (running == 0) {
+		/* only close the driver if initialized and
+		   this is the last worker thread */
+		if (!o->use_posixio && cufile_initialized) {
+			cuFileDriverClose();
+		}
+		cufile_initialized = 0;
+	}
+	pthread_mutex_unlock(&running_lock);
+}
+
+FIO_STATIC struct ioengine_ops ioengine = {
+	.name                = "cufile",
+	.version             = FIO_IOOPS_VERSION,
+	.init                = fio_cufile_init,
+	.queue               = fio_cufile_queue,
+	.open_file           = fio_cufile_open_file,
+	.close_file          = fio_cufile_close_file,
+	.iomem_alloc         = fio_cufile_iomem_alloc,
+	.iomem_free          = fio_cufile_iomem_free,
+	.cleanup             = fio_cufile_cleanup,
+	.flags               = FIO_SYNCIO,
+	.options             = options,
+	.option_struct_size  = sizeof(struct cufile_options)
+};
+
+void fio_init fio_cufile_register(void)
+{
+	const char* diskless = getenv("FIO_CUFILE_DISKLESSIO");
```

Comment:
Hmm this is different. We don't typically pass parameters via environment variable... What forced you to do so here?

Reply:
Agreed, that is out of the ordinary. The reason for using an environment variable is because the underlying filesystem may or may not be diskless. Using an fio_option to set the flag didn't appear to work. I was also unclear on the total ramifications of setting FIO_DISKLESSIO. I see it used in engines/cpu.c and engines/rdma.c, which makes perfect sense, but also in engines/glusterfs_sync.c, and that made things murkier. My current take is that FIO_DISKLESSIO should be set if the engine *should never* access local disk.

Long story short, I don't believe that FIO_DISKLESSIO needs to be set and will remove that code.